src/DiffSharp.Core/Model.fs

```diff
@@ -792,4 +808,200 @@ type VAE(xDim:int, zDim:int, ?hDims:seq<int>, ?nonlinearity:Tensor->Tensor, ?non
     /// <summary>TBD</summary>
     member _.sample(?numSamples:int) = 
         let numSamples = defaultArg numSamples 1
-        dsharp.randn([|numSamples; zDim|]) |> decode
\ No newline at end of file
+        dsharp.randn([|numSamples; zDim|]) |> decode
+
+
+
+/// <summary>Unit cell of a recurrent neural network. Prefer using the RNN class instead, which can combine RNNCells in multiple layers.</summary>
+type RNNCell(inFeatures, outFeatures, ?nonlinearity, ?bias, ?batchFirst) =
+    inherit Model()
+    let nonlinearity = defaultArg nonlinearity dsharp.tanh
+    let bias = defaultArg bias true
+    let batchFirst = defaultArg batchFirst false
+    let k = 1./sqrt (float outFeatures)
+    let wih = Parameter(Weight.uniform([|inFeatures; outFeatures|], k))
+    let whh = Parameter(Weight.uniform([|outFeatures; outFeatures|], k))
+    let b = Parameter(if bias then Weight.uniform([|outFeatures|], k) else dsharp.tensor([]))
+    let h = Parameter <| dsharp.tensor([]) // Not a paramter to be trained, this is for keeping hidden state
+    do base.add([wih;whh;b],["RNNCell-weight-ih";"RNNCell-weight-hh";"RNNCell-bias"])
+
+    member _.hidden 
+        with get () = h.value
+        and set v = h.value <- v
+
+    override _.getString() = sprintf "RNNCell(%A, %A)" inFeatures outFeatures
+
+    member r.reset() = r.hidden <- dsharp.tensor([])
+
+    override r.forward(value) =
+        let value, seqLen, batchSize = rnnShape value inFeatures batchFirst
+        if r.hidden.nelement = 0 then r.hidden <- dsharp.zeros([batchSize; outFeatures])
+        let output = Array.create seqLen (dsharp.tensor([]))
+        for i in 0..seqLen-1 do
+            let v = value.[i]
+            r.hidden <- dsharp.matmul(v, wih.value) + dsharp.matmul(h.value, whh.value)
+            if bias then r.hidden <- r.hidden + b.value
+            r.hidden <- nonlinearity r.hidden
+            output.[i] <- r.hidden
+        let output = dsharp.stack output
+        if batchFirst then output.transpose(0, 1) else output
+
+/// <summary>Unit cell of a long short-term memory (LSTM) recurrent neural network. Prefer using the RNN class instead, which can combine RNNCells in multiple layers.</summary>
+type LSTMCell(inFeatures, outFeatures, ?bias, ?batchFirst) =
```

Comment:
I recommend we should move each actual model out into a separate file.  This file is getting too long

Reply:
Agreed